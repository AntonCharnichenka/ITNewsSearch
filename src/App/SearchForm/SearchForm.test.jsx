import  { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import { SearchForm } from './index';

describe(
    'SearchForm',
    () => {
        const searchFormProps = {
            searchTerm: 'React',
            onSearchInput: vi.fn(),
            onSearchSubmit: vi.fn(),
        };

        it(
            'renders input field with its value',
            () => {
                render(<SearchForm {...searchFormProps}/>);
                // screen.debug();
                expect(screen.getByDisplayValue('React')).toBeInTheDocument();
            }
        );

        it(
            'renders correct label',
            () => {
                render(<SearchForm {...searchFormProps}/>);
                expect(screen.getByLabelText(/Search/)).toBeInTheDocument();
            }
        );

        it(
            'calls onSearchInput on input field change',
            () => {
                render(<SearchForm {...searchFormProps}/>);
                
                fireEvent.change(
                    screen.getByDisplayValue('React'),
                    { target: { value: 'Redux' }}
                );

                expect(searchFormProps.onSearchInput).toHaveBeenCalledTimes(1);
                // expect(searchFormProps.onSearchInput).toHaveBeenCalledWith(
                //     expect.objectContaining(
                //         { target: expect.objectContaining({ value: 'Redux' }) }
                //     )
                // );
            }
        );

        it(
            'calls onSearchSubmit on button submit click',
            () => {
                render(<SearchForm {...searchFormProps}/>);
                
                fireEvent.submit(screen.getByRole('button'));

                expect(searchFormProps.onSearchSubmit).toHaveBeenCalledTimes(1);
            }
        )

        it(
            'renders snapshot',
            () => {
                const { container } = render(<SearchForm {...searchFormProps}/>);
                expect(container.firstChild).toMatchSnapshot();

                // Snapshot testing does not inherently capture the runtime state of elements, 
                // such as whether an input is focused or not. 
                // The focus state is a dynamic DOM state that isn't represented directly in the markup or static attributes that would be visible in a snapshot.

                // Focus is not a DOM attribute: It does not have a corresponding attribute in the HTML output that snapshot testing would capture.
                // Focus is managed by the browser: The focus state is managed by the browser and typically only affects the styling and behavior (such as which element responds to keyboard inputs), rather than the underlying HTML.
            }
        )
    }
);
