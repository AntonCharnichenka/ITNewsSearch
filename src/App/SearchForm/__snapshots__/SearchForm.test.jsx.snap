// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchForm > renders snapshot 1`] = `
<form
  class="sc-iGgVNO fgzkTm"
>
  <label
    class="sc-aYaIB iJLwbK"
    for="search"
  >
    <strong>
      Search:
    </strong>
  </label>
  Â 
  <input
    class="sc-gEvDqW ggpFwF"
    id="search"
    type="text"
    value="React"
  />
  <button
    class="sc-eqUzNf sc-fqkwJk chkPTJ hZzotC"
    type="submit"
  >
    <svg
      fill="currentColor"
      height="1em"
      id="search"
      stroke="currentColor"
      stroke-width="0"
      version="1.1"
      viewBox="0 0 24 24"
      width="1em"
      x="0px"
      xmlns="http://www.w3.org/2000/svg"
      y="0px"
    >
      <g>
        <path
          d="M20.031,20.79c0.46,0.46,1.17-0.25,0.71-0.7l-3.75-3.76c1.27-1.41,2.04-3.27,2.04-5.31
		c0-4.39-3.57-7.96-7.96-7.96s-7.96,3.57-7.96,7.96c0,4.39,3.57,7.96,7.96,7.96c1.98,0,3.81-0.73,5.21-1.94L20.031,20.79z
		 M4.11,11.02c0-3.84,3.13-6.96,6.96-6.96c3.84,0,6.96,3.12,6.96,6.96c0,3.84-3.12,6.96-6.96,6.96C7.24,17.98,4.11,14.86,4.11,11.02
		z"
        />
      </g>
    </svg>
  </button>
</form>
`;
